### PR TITLE
[V0182] Add /presidential/contributions/by_candidate/

### LINKED ISSUE #4170: Make endpoint for candidate list (presidential/contributions/by_candidate)
### Summary
Design ticket: https://github.com/fecgov/openFEC/issues/4175
Copy of Endpoints for presidential bubble map-Presidential bubble map

Notes: 
### 1) /presidential/contributions/by_candidate/
combine state data with national data

TODO- Add 2016 tables
(left side cand list, national and state wide get all candidate list)
(view on DISCLOSURE.PRES_CA_CM_SCHED_STATE_20D / _16 - state data
UNION PRES_F3P_TOTALS_CA_CM_LINK_20D/ _16 - national data)
Filter: contributor_state, election_year

Include rounded and original totals
```
--US totals
SELECT DISTINCT
cand_id AS candidate_id, 
--TRIM(TRAILING ',' FROM cand_nm) AS candidate_name, --remove trailing commas
SUBSTR(cand_nm, 0, strpos(cand_nm,',')) AS candidate_last_name, 
(CASE WHEN cand_pty_affiliation IS NULL THEN NULL ELSE '('||SUBSTR(CAND_PTY_AFFILIATION,1,1)||')' END) AS candidate_party_affiliation, 
((INDV_CONTB_PER - REF_INDV_CONTB_PER) + 
    (POL_PTY_CMTE_CONTB_PER - REF_POL_PTY_CMTE_CONTB_PER) + 
    (OTHER_POL_CMTE_CONTB_PER - REF_OTHER_POL_CMTE_CONTB_PER) + 
    CAND_CONTB_PER + TRANF_FROM_AFFILATED_CMTE_PER + 
    (LOANS_RECEIVED_FROM_CAND_PER - REPYMTS_LOANS_MADE_BY_CAND_PER) + 
    (OTHER_LOANS_RECEIVED_PER - REPYMTS_OTHER_LOANS_PER) + OTHER_RECEIPTS_PER 
    + COALESCE(FED_FUNDS_PER,0)) AS net_receipts,
ROUND(((INDV_CONTB_PER - REF_INDV_CONTB_PER) + 
    (POL_PTY_CMTE_CONTB_PER - REF_POL_PTY_CMTE_CONTB_PER) + 
    (OTHER_POL_CMTE_CONTB_PER - REF_OTHER_POL_CMTE_CONTB_PER) + 
    CAND_CONTB_PER + TRANF_FROM_AFFILATED_CMTE_PER + 
    (LOANS_RECEIVED_FROM_CAND_PER - REPYMTS_LOANS_MADE_BY_CAND_PER) + 
    (OTHER_LOANS_RECEIVED_PER - REPYMTS_OTHER_LOANS_PER) + OTHER_RECEIPTS_PER 
    + COALESCE(FED_FUNDS_PER,0))/1000000,1) AS rounded_net_receipts,
'US' AS contributor_state,
2020 as election_year -- TODO: Bring in other 2016 data
FROM DISCLOSURE.PRES_F3P_TOTALS_CA_CM_LINK_20D
UNION
-- Per-state totals
SELECT CAND_ID as candidate_id,
CAND_NM  as candidate_last_name, 
CAND_PTY_AFFILIATION as candidate_party_affiliation, 
SUM(NET_RECEIPTS_STATE) AS net_receipts,
ROUND (SUM(NET_RECEIPTS_STATE )/1000000,1) as rounded_net_receipts, --per state
contbr_st as contributor_state,
2020 as election_year -- TODO: Bring in other 2016 data
FROM DISCLOSURE.PRES_CA_CM_SCHED_STATE_20D
GROUP BY candidate_id,candidate_party_affiliation,candidate_last_name, contributor_state
ORDER BY contributor_state, rounded_net_receipts DESC
```
### Completion criteria
- [ ] Make views in database with migration (pair w/DB team)
- [ ] Make endpoints based on design
- [ ] Use a feature flag
- [ ] Add test coverage for new endpoints

### PR DESCRIPTION
**Pull request by lbeaufort**
## Summary (required)

- Resolves #4170 

Add `/presidential/contributions/by_candidate/`
Future work: Will check with @fecjjeng on indexes as part of #4186 performance testing

## How to test the changes locally

- `export FEC_FEATURE_PRESIDENTIAL=True`
- Change `PresidentialByCandidate` tablename to `'ofec_presidential_by_candidate_vw_lb'`
- Test a query with Swagger
- http://localhost:5000/v1/presidential/contributions/by_candidate/?contributor_state=US&election_year=2020&per_page=100&page=1&sort_hide_null=false&sort_null_only=false&sort_nulls_last=false
- Test migration with 
```
dropdb cfdm_test
createdb cfdm_test
invoke create_sample_db
flyway migrate
```

## Impacted areas of the application
List general components of the application that this PR will affect:

-  Future presidential map


**Query plan:**
```
explain(SELECT public.ofec_presidential_by_candidate_vw_lb.idx, public.ofec_presidential_by_candidate_vw_lb.candidate_id, public.ofec_presidential_by_candidate_vw_lb.candidate_last_name, public.ofec_presidential_by_candidate_vw_lb.candidate_party_affiliation, public.ofec_presidential_by_candidate_vw_lb.net_receipts, public.ofec_presidential_by_candidate_vw_lb.rounded_net_receipts, public.ofec_presidential_by_candidate_vw_lb.contributor_state, public.ofec_presidential_by_candidate_vw_lb.election_year  
 FROM public.ofec_presidential_by_candidate_vw_lb  
 WHERE public.ofec_presidential_by_candidate_vw_lb.contributor_state = 'US' AND public.ofec_presidential_by_candidate_vw_lb.election_year = 2020 ORDER BY public.ofec_presidential_by_candidate_vw_lb.net_receipts DESC)    
```
```                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| QUERY PLAN                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Sort  (cost=1350.06..1350.06 rows=1 width=208)                                                                                                                                                                                                                                                                                                                                                                                                                                                                       |
|   Sort Key: ofec_presidential_by_candidate_vw_lb.net_receipts DESC                                                                                                                                                                                                                                                                                                                                                                                                                                                   |
|   ->  Subquery Scan on ofec_presidential_by_candidate_vw_lb  (cost=1321.33..1350.05 rows=1 width=208)                                                                                                                                                                                                                                                                                                                                                                                                                |
|         Filter: ((ofec_presidential_by_candidate_vw_lb.contributor_state = 'US'::text) AND (ofec_presidential_by_candidate_vw_lb.election_year = 2020))                                                                                                                                                                                                                                                                                                                                                              |
|         ->  WindowAgg  (cost=1321.33..1339.28 rows=718 width=208)                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
|               ->  Sort  (cost=1321.33..1323.12 rows=718 width=200)                                                                                                                                                                                                                                                                                                                                                                                                                                                   |
|                     Sort Key: (2020), ('US'::text), ((((((((((pres_f3p_totals_ca_cm_link_20d.indv_contb_per - pres_f3p_totals_ca_cm_link_20d.ref_indv_contb_per) + (pres_f3p_totals_ca_cm_link_20d.pol_pty_cmte_contb_per - pres_f3p_totals_ca_cm_link_20d.ref_pol_pty_cmte_contb_per)) + (pres_f3p_totals_ca_cm_link_20d.other_pol_cmte_contb_per - pres_f3p_totals_ca_cm_link_20d.ref_other_pol_cmte_contb_per)) + pres_f3p_totals_ca_cm_link_20d.cand_contb_per) + pres_f3p_totals_ca_cm_link_20d.tranf_from_a... |
|                     ->  HashAggregate  (cost=1280.08..1287.26 rows=718 width=200)                                                                                                                                                                                                                                                                                                                                                                                                                                    |
|                           Group Key: pres_f3p_totals_ca_cm_link_20d.cand_id, (substr((pres_f3p_totals_ca_cm_link_20d.cand_nm)::text, 0, strpos((pres_f3p_totals_ca_cm_link_20d.cand_nm)::text, ','::text))), pres_f3p_totals_ca_cm_link_20d.cand_pty_affiliation, ((((((((((pres_f3p_totals_ca_cm_link_20d.indv_contb_per - pres_f3p_totals_ca_cm_link_20d.ref_indv_contb_per) + (pres_f3p_totals_ca_cm_link_20d.pol_pty_cmte_contb_per - pres_f3p_totals_ca_cm_link_20d.ref_pol_pty_cmte_contb_per)) + (pres_f3p... |
|                           ->  Append  (cost=30.75..1267.52 rows=718 width=200)                                                                                                                                                                                                                                                                                                                                                                                                                                       |
|                                 ->  Unique  (cost=30.75..33.30 rows=170 width=146)                                                                                                                                                                                                                                                                                                                                                                                                                                   |
|                                       ->  Sort  (cost=30.75..31.17 rows=170 width=146)                                                                                                                                                                                                                                                                                                                                                                                                                               |
|                                             Sort Key: pres_f3p_totals_ca_cm_link_20d.cand_id, (substr((pres_f3p_totals_ca_cm_link_20d.cand_nm)::text, 0, strpos((pres_f3p_totals_ca_cm_link_20d.cand_nm)::text, ','::text))), pres_f3p_totals_ca_cm_link_20d.cand_pty_affiliation, ((((((((((pres_f3p_totals_ca_cm_link_20d.indv_contb_per - pres_f3p_totals_ca_cm_link_20d.ref_indv_contb_per) + (pres_f3p_totals_ca_cm_link_20d.pol_pty_cmte_contb_per - pres_f3p_totals_ca_cm_link_20d.ref_pol_pty_cmte_contb_... |
|                                             ->  Seq Scan on pres_f3p_totals_ca_cm_link_20d  (cost=0.00..24.45 rows=170 width=146)                                                                                                                                                                                                                                                                                                                                                                                    |
|                                 ->  Subquery Scan on "*SELECT* 2"  (cost=608.15..613.49 rows=194 width=174)                                                                                                                                                                                                                                                                                                                                                                                                          |
|                                       ->  Unique  (cost=608.15..611.55 rows=194 width=121)                                                                                                                                                                                                                                                                                                                                                                                                                           |
|                                             ->  Sort  (cost=608.15..608.64 rows=194 width=121)                                                                                                                                                                                                                                                                                                                                                                                                                       |
|                                                   Sort Key: pres_ca_cm_sched_state_20d.cand_id, pres_ca_cm_sched_state_20d.cand_nm, (expand_party_lb((pres_ca_cm_sched_state_20d.cand_pty_affiliation)::text)), (sum(pres_ca_cm_sched_state_20d.net_receipts_state)), (round((sum(pres_ca_cm_sched_state_20d.net_receipts_state) / '1000000'::numeric), 1)), pres_ca_cm_sched_state_20d.contbr_st                                                                                                                    |
|                                                   ->  HashAggregate  (cost=548.40..600.78 rows=194 width=121)                                                                                                                                                                                                                                                                                                                                                                                                        |
|                                                         Group Key: pres_ca_cm_sched_state_20d.cand_id, expand_party_lb((pres_ca_cm_sched_state_20d.cand_pty_affiliation)::text), pres_ca_cm_sched_state_20d.cand_nm, pres_ca_cm_sched_state_20d.contbr_st                                                                                                                                                                                                                                                            |
|                                                         ->  Seq Scan on pres_ca_cm_sched_state_20d  (cost=0.00..519.36 rows=1936 width=59)                                                                                                                                                                                                                                                                                                                                                                           |
|                                 ->  Unique  (cost=30.75..33.30 rows=170 width=146)                                                                                                                                                                                                                                                                                                                                                                                                                                   |
|                                       ->  Sort  (cost=30.75..31.17 rows=170 width=146)                                                                                                                                                                                                                                                                                                                                                                                                                               |
|                                             Sort Key: pres_f3p_totals_ca_cm_link_16.cand_id, (substr((pres_f3p_totals_ca_cm_link_16.cand_nm)::text, 0, strpos((pres_f3p_totals_ca_cm_link_16.cand_nm)::text, ','::text))), pres_f3p_totals_ca_cm_link_16.cand_pty_affiliation, ((((((((((pres_f3p_totals_ca_cm_link_16.indv_contb_per - pres_f3p_totals_ca_cm_link_16.ref_indv_contb_per) + (pres_f3p_totals_ca_cm_link_16.pol_pty_cmte_contb_per - pres_f3p_totals_ca_cm_link_16.ref_pol_pty_cmte_contb_per)) + ... |
|                                             ->  Seq Scan on pres_f3p_totals_ca_cm_link_16  (cost=0.00..24.45 rows=170 width=146)                                                                                                                                                                                                                                                                                                                                                                                     |
|                                 ->  Subquery Scan on "*SELECT* 4"  (cost=578.98..584.04 rows=184 width=174)                                                                                                                                                                                                                                                                                                                                                                                                          |
|                                       ->  Unique  (cost=578.98..582.20 rows=184 width=120)                                                                                                                                                                                                                                                                                                                                                                                                                           |
|                                             ->  Sort  (cost=578.98..579.44 rows=184 width=120)                                                                                                                                                                                                                                                                                                                                                                                                                       |
|                                                   Sort Key: pres_ca_cm_sched_state_16.cand_id, pres_ca_cm_sched_state_16.cand_nm, (expand_party_lb((pres_ca_cm_sched_state_16.cand_pty_affiliation)::text)), (sum(pres_ca_cm_sched_state_16.net_receipts_state)), (round((sum(pres_ca_cm_sched_state_16.net_receipts_state) / '1000000'::numeric), 1)), pres_ca_cm_sched_state_16.contbr_st                                                                                                                          |
|                                                   ->  HashAggregate  (cost=522.38..572.05 rows=184 width=120)                                                                                                                                                                                                                                                                                                                                                                                                        |
|                                                         Group Key: pres_ca_cm_sched_state_16.cand_id, expand_party_lb((pres_ca_cm_sched_state_16.cand_pty_affiliation)::text), pres_ca_cm_sched_state_16.cand_nm, pres_ca_cm_sched_state_16.contbr_st                                                                                                                                                                                                                                                                |
|                                                         ->  Seq Scan on pres_ca_cm_sched_state_16  (cost=0.00..494.70 rows=1845 width=58)                                                                                                                                                                                                                                                                                                                                                                            |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

### CODE REVIEW

File: data/migrations/V0182__add_ofec_presidential_by_candidate_vw.sql [hunk 1]
```diff
@@ -0,0 +1,110 @@
+/*
+
+View for https://github.com/fecgov/openFEC/issues/4170
+Adds presidential contributions by candidate, state and national totals
+Endpoint: /presidential/contributions/by_candidate/
+
```
Comment (lbeaufort) [inline data/migrations/V0182__add_ofec_presidential_by_candidate_vw.sql]: Specify schema?

**Comment by fecjjeng** [comment]
Downloaded the branch to local server and run the following tests:
migration file executed successfully.
pytest pass successfuly
local server execution passed.

**Comment by fecjjeng** [comment]
A quick question, do we need idx column in the view?

**Comment by fecjjeng** [comment]
Also, disclosure.pres_f3p_totals_ca_cm_link_20d is the form level of data and disclosure.pres_ca_cm_sched_state_20d is the summary of schedule level of data.  Not sure logically it should be together.  I had checked the lump sum of sched data and the form data per candidate, they are quite different.

**Comment by lbeaufort** [comment]
Thanks for your review, @fecjjeng! 

>A quick question, do we need idx column in the view?
SQLAlchemy needs at least one index column for a model. I have it hidden from the output
> Also, disclosure.pres_f3p_totals_ca_cm_link_20d is the form level of data and disclosure.pres_ca_cm_sched_state_20d is the summary of schedule level of data. Not sure logically it should be together. I had checked the lump sum of sched data and the form data per candidate, they are quite different.
I see what you're saying - doing a group by - the endpoint is to feed the national and state chart, which is why I combined them:
### Classic-style
Screen Shot 2020-02-05 at 3 34 22 PM
### Combined
Screen Shot 2020-02-05 at 3 36 55 PM

https://www.draw.io/#G108WMWi2-V0EOjRuKer8m0uzOu_aJK9M8

**Comment by lbeaufort** [comment]
I discussed this with @fecjjeng and @fec-jli. 

One endpoint is a bit messy from the data side, and two endpoints is a bit messy from the API side.

Because this solution as-is works, has test coverage, is blocking other development, and won't go live until we publish the presidential section, we're going to move forward with one endpoint, talk to @JonellaCulmer and @rfultz about what makes the front-end more straightforward, and then decide whether we should split this data into two endpoints before publishing it.

Thanks @fecjjeng!

**Pull request merged**